pep-0544.txt

```diff
@@ -0,0 +1,967 @@
+cases we propose to just add a qualifier such as *protocol classes*
+when referring to the static type concept.
+
+If a class includes a protocol in its MRO, the class is called
+an *explicit* subclass of the protocol. If a class is a structural subtype
+of a protocol, it is said to implement the protocol and to be compatible
+with a protocol. If a class is compatible with a protocol but the protocol
+is not included in the MRO, the class is an *implicit* subtype
+of the protocol.
+
+The attributes (variables and methods) of a protocol that are mandatory
+for other class in order to be considered a structural subtype are called
+protocol members.
+
+
+.. _definition:
+
+Defining a protocol
+-------------------
+
+Protocols are defined by including a special new class ``typing.Protocol``
+(an instance of ``abc.ABCMeta``) in the base classes list, preferably
+at the end of the list. Here is a simple example::
+
+  from typing import Protocol
+
+  class SupportsClose(Protocol):
+      def close(self) -> None:
+          ...
+
+Now if one defines a class ``Resource`` with a ``close()`` method that has
+a compatible signature, it would implicitly be a subtype of
+``SupportsClose``, since the structural subtyping is used for
+protocol types::
+
+  class Resource:
+      ...
+      def close(self) -> None:
+          self.file.close()
+          self.lock.release()
+
+Apart from few restrictions explicitly mentioned below, protocol types can
+be used in every context where a normal types can::
+
+  def close_all(things: Iterable[SupportsClose]) -> None:
+      for t in things:
+          t.close()
+
+  f = open('foo.txt')
+  r = Resource()
+  close_all([f, r])  # OK!
+  close_all([1])     # Error: 'int' has no 'close' method
+
+Note that both the user-defined class ``Resource`` and the built-in
+``IO`` type (the return type of ``open()``) are considered subtypes of
+``SupportsClose``, because they provide a ``close()`` method with
+a compatible type signature.
+
+
+Protocol members
+----------------
+
+All methods defined in the protocol class body are protocol members, both
+normal and decorated with ``@abstractmethod``. If some or all parameters of
+protocol method are not annotated, then their types are assumed to be ``Any``
+(see PEP 484). Bodies of protocol methods are type checked, except for methods
+decorated with ``@abstractmethod`` with trivial bodies. A trivial body can
+contain a docstring. Example::
+
+  from typing import Protocol
+  from abc import abstractmethod
+
+  class Example(Protocol):
+      def first(self) -> int:     # This is a protocol member
+          return 42
+
+      @abstractmethod
+      def second(self) -> int:    # Method without a default implementation
+          """Some method."""
+
+Note that although formally the implicit return type of a method with
+a trivial body is ``None``, type checker will not warn about above example,
+such convention is similar to how methods are defined in stub files.
```

Comment:
Hmm, I don't really like the idea of "trivial bodies". Remember that when explicitly subclassing an ABC, abstract methods can still be called, through `super()`. An abstract method that shouldn't be called ought to raise `NotImplementedError`.

In stubs, mypy already doesn't type-check trivial bodies (though I don't recall its exact definition of "trivial" -- it probably doesn't exactly match what you have here). If we want to specify that behavior for all checkers reading stubs (which would seem to make sense) it should go in PEP 484, not here.

Or perhaps I am missing a use case for these that you've found in real code?